test/integration/shadow_policy_integration_test.cc

```diff
@@ -0,0 +1,148 @@
+#include "test/integration/filters/repick_cluster_filter.h"
+#include "test/integration/http_integration.h"
+
+namespace Envoy {
+namespace {
+
+class ShadowPolicyIntegrationTest : public testing::TestWithParam<Network::Address::IpVersion>,
+                                    public HttpIntegrationTest {
+public:
+  ShadowPolicyIntegrationTest() : HttpIntegrationTest(Http::CodecClient::Type::HTTP2, GetParam()) {
+    setUpstreamCount(2);
+  }
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, ShadowPolicyIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+// Test request mirroring / shadowing with the cluster name in policy
+TEST_P(ShadowPolicyIntegrationTest, RequestMirrorPolicyWithCluster) {
+  config_helper_.addConfigModifier([](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    auto* cluster = bootstrap.mutable_static_resources()->add_clusters();
+    cluster->MergeFrom(bootstrap.static_resources().clusters()[0]);
+    cluster->set_name("cluster_1");
+    ConfigHelper::setHttp2(*cluster);
+  });
+
+  // Set the mirror policy with cluster name
+  config_helper_.addConfigModifier(
+      [&](envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager&
+              hcm) -> void {
+        auto* mirror_policy = hcm.mutable_route_config()
+                                  ->mutable_virtual_hosts(0)
+                                  ->mutable_routes(0)
+                                  ->mutable_route()
+                                  ->add_request_mirror_policies();
+        mirror_policy->set_cluster("cluster_1");
+      });
+
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  // Send header-only request
+  Http::TestRequestHeaderMapImpl request_headers{
+      {":method", "GET"}, {":path", "/test/long/url"}, {":scheme", "http"}, {":authority", "host"}};
+
+  IntegrationStreamDecoderPtr response =
+      sendRequestAndWaitForResponse(request_headers, 0, default_response_headers_, 0, 0);
+  EXPECT_TRUE(upstream_request_->complete());
+  EXPECT_EQ(0U, upstream_request_->bodyLength());
+  EXPECT_EQ("200", response->headers().getStatusValue());
+  EXPECT_EQ(0U, response->body().size());
+
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_cx_total")->value(), 1);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_0.upstream_cx_total")->value(), 1);
+
+  // Cleanup.
+  cleanupUpstreamAndDownstream();
+}
+
+// Test request mirroring / shadowing with the cluster header, but the value is not set in header.
+TEST_P(ShadowPolicyIntegrationTest, RequestMirrorPolicyWithClusterHeaderNoFilter) {
```

Comment:
Sorry, there is a typo from comments. This test is going to test the scenario that there is no cluster header in request headers.  Updated the test file.

Reply:
I was not referring to typo or naming. I just think `RequestMirrorPolicyWithoutClusterHeader` test itself doesn't provide too much value in terms of testing because this is expected failing behavior. 

I think we just need the integration test to verify the functional correctness. Unit test can be used to cover both positive and negative cases

Reply:
Okay, that's fine. The `getShadowCluster` will return empty string during routing in this test case. This integration test simulates the empty string scenario without any errors. The main request will still get the 200 response. We can delete it if you think it doesn't have too much value.

Reply:
Thank you! Done.